security/rbac/engine/engine.go

```diff
@@ -0,0 +1,187 @@
+// +build go1.10
+
+/*
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package engine
+
+import (
+	"strings"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
+	cel "github.com/google/cel-go/cel"
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+)
+
+// AuthorizationArgs is the input of CEL engine.
+type AuthorizationArgs struct {
+	md       metadata.MD
+	peerInfo *peer.Peer
+}
+
+// Decision is the enum type that represents different authorization
+// decisions a CEL engine can return.
+type Decision int32
+
+const (
+	// DecisionAllow indicates allowing the RPC to go through.
+	DecisionAllow Decision = iota
+	// DecisionDeny indicates denying the RPC from going through.
+	DecisionDeny
+	// DecisionUnknown indicates that there is insufficient information to
+	// determine whether or not an RPC call is authorized.
+	DecisionUnknown
+)
+
+// String returns the string representation of a Decision object.
+func (d Decision) String() string {
+	return [...]string{"DecisionAllow", "DecisionDeny", "DecisionUnknown"}[d]
+}
+
+// AuthorizationDecision is the output of CEL engine.
+type AuthorizationDecision struct {
+	decision   Decision
+	policyName string
```

Comment:
So here it would be a list of names.

Reply:
Currently I just concatenated the list of UNKNOWN policy names. Do you think I should just make this a list by itself, which would be length 1 for ALLOW/DENY decisions?

Reply:
I remembered seeing some discussions with Jiangtao that we prefer a list of string names instead of a concatenated string, for simplicity. I might get wrong. Can you please confirm with Jiangtao on this point?

Reply:
I thought about concatenation but after a second thought, I think a list of policy names will give better semantics. In the case of second authorization engine. The second authorization engine can ignore all the policies that has false evaluation and directly evaluate the unknown policies.

Reply:
Ok, let's do list of strings then. Is it possible for two authorization engines have the list of same policy names or something? Otherwise, how would it iterate over the same list of policies?

Reply:
Update: probably referring to third-party authorization engines, don't worry about it.